### PR TITLE
HypreSolver: a new interface for Hypre

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Add a new interface class for Hypre. The existing Hypre interface classes are all tightly coupled with MLMG, and are therefore not usable for linear systems not supported by MLMG. In that case, one can use this new class.

### CODE REVIEW

File: Tests/LinearSolvers/Hypre/MyTest.cpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        auto marker = [=] AMREX_GPU_DEVICE (int /*boxno*/, int i, int j, int k, int /*n*/)
+            -> bool
+        {
+            return nddom.strictly_contains(i,j,k);
+        };
+
+        // For variable n at (i,j,k) in Box boxno (local index), fill its row in
+        // the matrix.
+        // [in ] gid : gid[n] is the id for variable n at (i,j,k)
+        // [out] nols: # of columns in this row.
```
Comment (jmsexton03) [inline Tests/LinearSolvers/Hypre/MyTest.cpp]:
```suggestion
        // [out] ncols: # of columns in this row.
```

File: Tests/LinearSolvers/Hypre/MyTest.cpp [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+        auto peb = phi_eb;
+        auto pdom = phi_domain;
+        auto dlo = amrex::lbound(nddom);
+        auto dhi = amrex::ubound(nddom);
+
+        // For variable n at (i,j,k) in Box boxno (local index), fill its row in
+        // the matrix.
+        // [in ] gid : gid[n] is the id for variable n at (i,j,k)
+        // [out] nols: # of columns in this row.
```
Comment (jmsexton03) [inline Tests/LinearSolvers/Hypre/MyTest.cpp]:
```suggestion
        // [out] ncols: # of columns in this row.
```

File: Tests/LinearSolvers/Hypre/MyTest.cpp [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                    ++ncols;
+                } else {
+                    srhs -= s * pdom;
+                }
+            } else {
+                s0   -= s/hm;
+                srhs -= s/hm * peb;
+            }
+
+#if (AMREX_SPACEDIM == 3)
```
Comment (jmsexton03) [inline Tests/LinearSolvers/Hypre/MyTest.cpp]:
```suggestion
#if (AMREX_SPACEDIM > 2)
```

File: Src/Extern/HYPRE/AMReX_HypreSolver.H [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+    }
+
+    HYPRE_IJMatrix getA () const { return m_hypre_ij->A(); }
+    HYPRE_IJVector getb () const { return m_hypre_ij->b(); }
+    HYPRE_IJVector getx () const { return m_hypre_ij->x(); }
+
+public: // for cuda
+
+    template <class Marker>
+#ifdef AMREX_USE_CUDA
```
Comment (jmsexton03) [inline Src/Extern/HYPRE/AMReX_HypreSolver.H:53]:
```suggestion
#ifdef AMREX_USE_GPU
```
To be consistent with more broad ifdef below

File: Src/Extern/HYPRE/AMReX_HypreSolver.H [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+    m_A = m_hypre_ij->A();
+    m_b = m_hypre_ij->b();
+    m_x = m_hypre_ij->x();
+
+    fill_matrix(a_filler);
+}
+
+template <int MSS>
+template <class Marker>
+#ifdef AMREX_USE_CUDA
```
Comment (jmsexton03) [inline Src/Extern/HYPRE/AMReX_HypreSolver.H:224]:
```suggestion
#ifdef AMREX_USE_GPU
```
to be consistent with broadening this ifdef

**Review by jmsexton03 — APPROVED** [review]
My suggestion on ifdefs are primarily for consistency. It appears this new interface is for the 2D or 3D case, and is written to work with GPUs generally (not just CUDA)

File: Src/Extern/HYPRE/AMReX_HypreSolver.H (hunk 4 above)
Comment (WeiqunZhang) [reply Src/Extern/HYPRE/AMReX_HypreSolver.H:53]: The macro is a workaround for *nvcc* compiler.

File: Src/Extern/HYPRE/AMReX_HypreSolver.H (hunk 5 above)
Comment (WeiqunZhang) [reply Src/Extern/HYPRE/AMReX_HypreSolver.H:224]: It is meant for cuda compiler only.

**Pull request merged**